File: server/crypto_box.py
#!/usr/bin/env python3
"""
CLAUDEPAD — enkripsi lalu lintas.

Sebelum v3.0 seluruh perintah dikirim sebagai JSON polos di jaringan lokal:
siapa pun yang menyadap WiFi bisa membaca apa yang diketik. PIN hanya
melindungi autentikasi, bukan isi percakapan.

Rancangan sengaja dibuat sederhana dan tanpa dependensi luar:
  * Kunci sesi diturunkan dari PIN (atau token pairing) + nonce acak dari
    kedua pihak memakai PBKDF2-HMAC-SHA256. Kunci tidak pernah dikirim.
  * Setiap pesan dienkripsi AES-256-CTR bila `cryptography` tersedia;
    bila tidak, dipakai ChaCha20 murni-Python yang disertakan di sini.
  * Setiap pesan diberi tag HMAC-SHA256 (enkripsi lalu MAC) agar isi yang
    diubah di tengah jalan langsung ditolak.
  * Nomor urut mencegah pesan lama diputar ulang oleh penyadap.

Ini bukan pengganti TLS untuk internet terbuka, tetapi menutup penyadapan
dan pengubahan lalu lintas di jaringan lokal — yang memang lingkup CLAUDEPAD.
"""

import base64
import hashlib
import hmac
import os
import struct

try:
    from cryptography.hazmat.primitives.asymmetric import rsa as _rsa
    from cryptography.hazmat.primitives.asymmetric import padding as _asym_padding
    from cryptography.hazmat.primitives import hashes as _hashes
    from cryptography.hazmat.primitives import serialization as _serialization
    from cryptography.hazmat.backends import default_backend as _default_backend
    HAS_CRYPTOGRAPHY = True
except ImportError:
    HAS_CRYPTOGRAPHY = False

KEY_LEN = 32
NONCE_LEN = 16
TAG_LEN = 32
PBKDF2_ROUNDS = 60_000


# ----------------------------------------------------------- ChaCha20 ------
def _rotl(v, c):
    return ((v << c) & 0xFFFFFFFF) | (v >> (32 - c))


def _quarter(x, a, b, c, d):
    x[a] = (x[a] + x[b]) & 0xFFFFFFFF; x[d] ^= x[a]; x[d] = _rotl(x[d], 16)
    x[c] = (x[c] + x[d]) & 0xFFFFFFFF; x[b] ^= x[c]; x[b] = _rotl(x[b], 12)
    x[a] = (x[a] + x[b]) & 0xFFFFFFFF; x[d] ^= x[a]; x[d] = _rotl(x[d], 8)
    x[c] = (x[c] + x[d]) & 0xFFFFFFFF; x[b] ^= x[c]; x[b] = _rotl(x[b], 7)


def _chacha_block(key, counter, nonce):
    const = b"expand 32-byte k"
    state = list(struct.unpack("<4I", const)) \
        + list(struct.unpack("<8I", key)) \
        + [counter] + list(struct.unpack("<3I", nonce))
    work = state[:]
    for _ in range(10):
        _quarter(work, 0, 4, 8, 12); _quarter(work, 1, 5, 9, 13)
        _quarter(work, 2, 6, 10, 14); _quarter(work, 3, 7, 11, 15)
        _quarter(work, 0, 5, 10, 15); _quarter(work, 1, 6, 11, 12)
        _quarter(work, 2, 7, 8, 13); _quarter(work, 3, 4, 9, 14)
    out = [(work[i] + state[i]) & 0xFFFFFFFF for i in range(16)]
    return struct.pack("<16I", *out)


def chacha20(key, nonce12, data, counter=1):
    """Enkripsi/dekripsi (operasinya sama, XOR aliran kunci)."""
    out = bytearray(len(data))
    for i in range(0, len(data), 64):
        block = _chacha_block(key, counter + i // 64, nonce12)
        chunk = data[i:i + 64]
        for j, b in enumerate(chunk):
            out[i + j] = b ^ block[j]
    return bytes(out)


# ------------------------------------------------------------ Session ------
class Session:
    """Kunci sesi + enkripsi pesan dua arah."""

    def __init__(self, key: bytes):
        self.key = key
        self.enc_key = hashlib.sha256(key + b"|enc").digest()
        self.mac_key = hashlib.sha256(key + b"|mac").digest()
        self.send_seq = 0
        self.recv_seq = -1

    # -- turunan kunci --
    @staticmethod
    def derive(secret: str, salt: bytes) -> "Session":
        key = hashlib.pbkdf2_hmac("sha256", secret.encode("utf-8"),
                                  salt, PBKDF2_ROUNDS, KEY_LEN)
        return Session(key)

    # -- enkripsi --
    def seal(self, plaintext: bytes) -> bytes:
        """seq(8) | nonce(12) | ciphertext | tag(32)"""
        self.send_seq += 1
        seq = struct.pack(">Q", self.send_seq)
        nonce = os.urandom(12)
        ct = chacha20(self.enc_key, nonce, plaintext)
        body = seq + nonce + ct
        tag = hmac.new(self.mac_key, body, hashlib.sha256).digest()
        return body + tag

    def open(self, blob: bytes) -> bytes:
        if len(blob) < 8 + 12 + TAG_LEN:
            raise ValueError("paket terlalu pendek")
        body, tag = blob[:-TAG_LEN], blob[-TAG_LEN:]
        expect = hmac.new(self.mac_key, body, hashlib.sha256).digest()
        if not hmac.compare_digest(tag, expect):
            raise ValueError("tag tidak cocok — isi pesan diubah di tengah jalan")
        seq = struct.unpack(">Q", body[:8])[0]
        if seq <= self.recv_seq:
            raise ValueError("pesan lama diputar ulang")
        self.recv_seq = seq
        nonce = body[8:20]
        return chacha20(self.enc_key, nonce, body[20:])


def new_salt() -> bytes:
    return os.urandom(NONCE_LEN)


# -------------------------------------------------------- RSA-2048 ---------
def generate_rsa_keypair():
    """Generate RSA-2048 keypair. Returns (public_key_pem, private_key).

    Dipanggil sekali saat server start. Public key dikirim ke HP di hello_ok,
    private key dipakai server untuk mendekripsi PIN/token dari HP.
    """
    if not HAS_CRYPTOGRAPHY:
        raise RuntimeError("modul 'cryptography' belum terpasang — jalankan: pip install cryptography")
    private_key = _rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
        backend=_default_backend(),
    )
    pub_pem = private_key.public_key().public_bytes(
        encoding=_serialization.Encoding.PEM,
        format=_serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return pub_pem, private_key


def rsa_encrypt(pub_pem: bytes, data: bytes) -> bytes:
    """Enkripsi data dengan RSA public key (OAEP + SHA-256)."""
    public_key = _serialization.load_pem_public_key(pub_pem, backend=_default_backend())
    return public_key.encrypt(
        data,
        _asym_padding.OAEP(
            mgf=_asym_padding.MGF1(algorithm=_hashes.SHA256()),
            algorithm=_hashes.SHA256(),
            label=None,
        ),
    )


def rsa_decrypt(private_key, encrypted: bytes) -> bytes:
    """Dekripsi data dengan RSA private key (OAEP + SHA-256)."""
    return private_key.decrypt(
        encrypted,
        _asym_padding.OAEP(
            mgf=_asym_padding.MGF1(algorithm=_hashes.SHA256()),
            algorithm=_hashes.SHA256(),
            label=None,
        ),
    )


def rsa_pubkey_to_b64(pub_pem: bytes) -> str:
    """Konversi public key PEM ke base64 (tanpa header PEM)."""
    lines = pub_pem.decode("ascii").splitlines()
    raw = "".join(l for l in lines if not l.startswith("-----"))
    return raw


def rsa_pubkey_from_b64(b64: str) -> bytes:
    """Konversi base64 kembali ke public key PEM."""
    header = "-----BEGIN PUBLIC KEY-----"
    footer = "-----END PUBLIC KEY-----"
    raw = b64.strip()
    return (header + "\n" + raw + "\n" + footer).encode("ascii")

File: android/app/src/main/java/com/bjorn/claudepad/CryptoBox.kt
package com.bjorn.claudepad

import android.util.Base64
import java.security.KeyFactory
import java.security.SecureRandom
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.PSource
import javax.crypto.spec.SecretKeySpec
import java.security.spec.MGF1ParameterSpec

/**
 * Enkripsi lalu lintas — pasangan Kotlin dari server/crypto_box.py.
 *
 * Sebelum v3.0 semua perintah dikirim JSON polos: penyadap WiFi bisa membaca
 * apa yang diketik. Kini kunci sesi diturunkan dari PIN/token + garam acak
 * memakai PBKDF2, dan tidak pernah dikirim. Tiap pesan dienkripsi ChaCha20
 * lalu diberi tag HMAC-SHA256 (encrypt-then-MAC) dengan nomor urut anti-ulang.
 *
 * Harus identik byte-per-byte dengan sisi Python, karena keduanya menurunkan
 * kunci sendiri dari rahasia yang sama.
 */
class CryptoBox private constructor(key: ByteArray) {

    private val encKey = sha256(key + "|enc".toByteArray())
    private val macKey = sha256(key + "|mac".toByteArray())
    private var sendSeq = 0L
    private var recvSeq = -1L
    private val rnd = SecureRandom()

    companion object {
        private const val PBKDF2_ROUNDS = 60_000
        private const val KEY_LEN = 32
        const val TAG_LEN = 32

        fun derive(secret: String, salt: ByteArray): CryptoBox {
            val spec = PBEKeySpec(secret.toCharArray(), salt, PBKDF2_ROUNDS, KEY_LEN * 8)
            val key = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
                .generateSecret(spec).encoded
            return CryptoBox(key)
        }

        /**
         * Enkripsi data dengan RSA public key (OAEP + SHA-256).
         * Dipakai untuk mengenkripsi PIN/token sebelum dikirim ke server.
         * pubKeyBase64: public key dalam format base64 (tanpa header PEM).
         */
        fun encryptWithPublicKey(pubKeyBase64: String, data: ByteArray): ByteArray {
            val keyBytes = Base64.decode(pubKeyBase64, Base64.DEFAULT)
            val keyFactory = KeyFactory.getInstance("RSA")
            val spec = X509EncodedKeySpec(keyBytes)
            val publicKey = keyFactory.generatePublic(spec)
            val cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding")
            cipher.init(
                Cipher.ENCRYPT_MODE, publicKey,
                OAEPParameterSpec(
                    "SHA-256", "MGF1",
                    MGF1ParameterSpec("SHA-256"),
                    PSource.PSpecified(ByteArray(0))
                )
            )
            return cipher.doFinal(data)
        }

        private fun sha256(b: ByteArray): ByteArray =
            java.security.MessageDigest.getInstance("SHA-256").digest(b)
    }

    /** seq(8) | nonce(12) | ciphertext | tag(32) */
    fun seal(plain: ByteArray): ByteArray {
        sendSeq++
        val seq = longToBytes(sendSeq)
        val nonce = ByteArray(12).also { rnd.nextBytes(it) }
        val ct = chacha20(encKey, nonce, plain, 1)
        val body = seq + nonce + ct
        return body + hmac(macKey, body)
    }

    fun open(blob: ByteArray): ByteArray {
        if (blob.size < 8 + 12 + TAG_LEN) throw SecurityException("paket terlalu pendek")
        val body = blob.copyOfRange(0, blob.size - TAG_LEN)
        val tag = blob.copyOfRange(blob.size - TAG_LEN, blob.size)
        if (!constantEquals(tag, hmac(macKey, body)))
            throw SecurityException("tag tidak cocok")
        val seq = bytesToLong(body, 0)
        if (seq <= recvSeq) throw SecurityException("pesan lama diputar ulang")
        recvSeq = seq
        val nonce = body.copyOfRange(8, 20)
        return chacha20(encKey, nonce, body.copyOfRange(20, body.size), 1)
    }

    // ---------------------------------------------------------- ChaCha20 --
    private fun chacha20(key: ByteArray, nonce: ByteArray, data: ByteArray,
                         counter: Int): ByteArray {
        val out = ByteArray(data.size)
        var i = 0
        while (i < data.size) {
            val block = chachaBlock(key, counter + i / 64, nonce)
            var j = 0
            while (j < 64 && i + j < data.size) {
                out[i + j] = (data[i + j].toInt() xor block[j].toInt()).toByte()
                j++
            }
            i += 64
        }
        return out
    }

    private fun rotl(v: Int, c: Int) = (v shl c) or (v ushr (32 - c))

    private fun chachaBlock(key: ByteArray, counter: Int, nonce: ByteArray): ByteArray {
        val state = IntArray(16)
        state[0] = 0x61707865; state[1] = 0x3320646e
        state[2] = 0x79622d32; state[3] = 0x6b206574
        for (k in 0 until 8) state[4 + k] = leInt(key, k * 4)
        state[12] = counter
        for (k in 0 until 3) state[13 + k] = leInt(nonce, k * 4)
        val w = state.copyOf()
        repeat(10) {
            qr(w, 0, 4, 8, 12); qr(w, 1, 5, 9, 13)
            qr(w, 2, 6, 10, 14); qr(w, 3, 7, 11, 15)
            qr(w, 0, 5, 10, 15); qr(w, 1, 6, 11, 12)
            qr(w, 2, 7, 8, 13); qr(w, 3, 4, 9, 14)
        }
        val out = ByteArray(64)
        for (k in 0 until 16) {
            val v = w[k] + state[k]
            out[k * 4] = v.toByte()
            out[k * 4 + 1] = (v ushr 8).toByte()
            out[k * 4 + 2] = (v ushr 16).toByte()
            out[k * 4 + 3] = (v ushr 24).toByte()
        }
        return out
    }

    private fun qr(x: IntArray, a: Int, b: Int, c: Int, d: Int) {
        x[a] += x[b]; x[d] = rotl(x[d] xor x[a], 16)
        x[c] += x[d]; x[b] = rotl(x[b] xor x[c], 12)
        x[a] += x[b]; x[d] = rotl(x[d] xor x[a], 8)
        x[c] += x[d]; x[b] = rotl(x[b] xor x[c], 7)
    }

    private fun leInt(b: ByteArray, off: Int): Int =
        (b[off].toInt() and 0xFF) or
        ((b[off + 1].toInt() and 0xFF) shl 8) or
        ((b[off + 2].toInt() and 0xFF) shl 16) or
        ((b[off + 3].toInt() and 0xFF) shl 24)

    // ---------------------------------------------------------- helpers ---
    private fun hmac(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun constantEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var r = 0
        for (i in a.indices) r = r or (a[i].toInt() xor b[i].toInt())
        return r == 0
    }

    private fun longToBytes(v: Long): ByteArray {
        val o = ByteArray(8)
        for (i in 0 until 8) o[i] = (v ushr (56 - i * 8)).toByte()
        return o
    }

    private fun bytesToLong(b: ByteArray, off: Int): Long {
        var v = 0L
        for (i in 0 until 8) v = (v shl 8) or (b[off + i].toLong() and 0xFF)
        return v
    }
}
